node/impl/full/eth.go

```diff
@@ -980,25 +980,32 @@ func (a *EthModule) applyMessage(ctx context.Context, msg *types.Message, tsk ty
 		return nil, xerrors.Errorf("cannot get tipset: %w", err)
 	}
 
-	applyTsMessages := true
-	if os.Getenv("LOTUS_SKIP_APPLY_TS_MESSAGE_CALL_WITH_GAS") == "1" {
-		applyTsMessages = false
+	// Walk back until we find a tipset where no state migration is needed
+	if ts.Height() > 0 {
+		for {
+			pts, err := a.Chain.GetTipSetFromKey(ctx, ts.Parents())
+			if err != nil {
+				return nil, xerrors.Errorf("failed to find a non-forking epoch: %w", err)
+			}
+			if !a.StateManager.HasExpensiveForkBetween(pts.Height(), ts.Height()+1) {
+				break
+			}
+			ts, err = a.Chain.GetTipSetFromKey(ctx, ts.Parents())
+			if err != nil {
+				return nil, xerrors.Errorf("getting parent tipset: %w", err)
+			}
+		}
 	}
```

Comment:
Actually I think the correct thing to do is not loop because the request is for an explicit tipset, so we can't just adjust the tipset they want. I believe the right behaviour is to just check whether there's an expensive fork and error if not. Perhaps like this:

```suggestion
	if ts.Height() > 0 {
		pts, err := a.Chain.GetTipSetFromKey(ctx, ts.Parents())
		if err != nil {
			return nil, xerrors.Errorf("failed to find a non-forking epoch: %w", err)
		}
		// Check for expensive forks from the parents to the tipset, including nil tipsets
		if a.StateManager.HasExpensiveForkBetween(pts.Height(), ts.Height()+1) {
			return nil, stmgr.ErrExpensiveFork
		}
	}
```

Then the call just won't work for those major upgrade epochs. I don't know if that's going to cause complaints but it's only a few every year.

Reply:
@rvagg But the pre-existing behaviour does loop back till it can find an appropriate tipset. I am not sure we should change that behaviour.

Reply:
only if `ts` is not supplied; but I think here we always have a `ts` so we're in the `else` part of this block: https://github.com/filecoin-project/lotus/blob/f75bdd1f262483ec13e82603687f7cf8b9d893b8/chain/stmgr/call.go#L104-L132

Reply:
@aarshkshah1992 ping on this; I think we should be able to get this wrapped up after dealing with this as long as we can get an expert reviewer to eyeball it.

Reply:
I think we agree that the loop can go, there's no need to "walk". The `if `ts.Height()` can go too as per @raulk's comment of there always being a height, the remaining concern here is about expensive forks and whether to do anything about them.

@raulk the current code path for this hits that check already, the reason for inserting it in here is that the new code bypasses it by taking shortcuts to the execution. This block was intended to restore the existing behaviour of returning an `ErrExpensiveFork` if you happen to hit an upgrade epoch with the call.

Is that an unreasonable thing to do for this call? Would there be a security concern if a public API allowed users to make this call repeatedly across that boundary? tbh I'm not sure of the implications of any of this other than the existing code seems to go to some effort to ensure that you don't try and load state for a manual execution where that might involve a big migration.

Reply:
@rvagg 

- I see your point about removing the loop. Yeah, we always have a tipset present (is resolved to the head if none is given by the user, same for pending). Fixed this.